src/engine/source/builder/builders/opBuilderSHAfrom.cpp

```diff
@@ -0,0 +1,123 @@
+#include "opBuilderSHAfrom.hpp"
+
+#include <optional>
+#include <string>
+
+#include <fmt/format.h>
+#include <json/json.hpp>
+
+#include <openssl/evp.h>
+#include <openssl/sha.h>
+
+#include "baseTypes.hpp"
+#include <baseHelper.hpp>
+
+namespace
+{
+
+// Sha1 digest len (20) * 2 (hex chars per byte)
+constexpr int OS_SHA1_HEXDIGEST_SIZE = (SHA_DIGEST_LENGTH * 2);
+constexpr int OS_SHA1_ARRAY_SIZE_LEN = OS_SHA1_HEXDIGEST_SIZE + 1;
+
+std::optional<std::string> wdbStringsHash(const std::vector<std::string>& input)
+{
+    char* parameter = NULL;
+    unsigned char digest[EVP_MAX_MD_SIZE];
+    unsigned int digest_size;
+
+    EVP_MD_CTX* ctx = EVP_MD_CTX_create();
+    if (!ctx)
+    {
+        // Failed during hash context creation
+        return std::nullopt;
+    }
+
+    if (1 != EVP_DigestInit_ex(ctx, EVP_sha1(), NULL))
+    {
+        // Failed during hash context initialization
+        EVP_MD_CTX_destroy(ctx);
+        return std::nullopt;
+    }
+
+    for (const auto& word : input)
+    {
+        if (1 != EVP_DigestUpdate(ctx, word.c_str(), word.length()))
+        {
+            // Failed during hash context update
+            return std::nullopt;
+        }
+    }
+
+    EVP_DigestFinal_ex(ctx, digest, &digest_size);
+    EVP_MD_CTX_destroy(ctx);
+
+    // OS_SHA1_Hexdigest(digest, hexdigest);
+    char output[OS_SHA1_ARRAY_SIZE_LEN];
+    for (size_t n = 0; n < SHA_DIGEST_LENGTH; n++)
+    {
+        sprintf(&output[n * 2], "%02x", digest[n]);
+    }
+
+    return {output};
+}
+
+} // namespace
+
+namespace builder::internals::builders
+{
+
+// field: +sha1_from/<string1>|<string_reference1>/<string2>|<string_reference2>
+base::Expression opBuilderSHAfrom(const std::any& definition)
```

Comment:
Please, improve the helper function and method names, and the file names and tests as well.

Reply:
Done!